src/Distributed/distributed_shallow_water_model.jl

```diff
@@ -36,7 +37,6 @@ function DistributedShallowWaterModel(; architecture, grid, boundary_conditions=
     my_model = ShallowWaterModel(;
                architecture = architecture,
                        grid = my_grid,
-    #    boundary_conditions = communicative_bcs,
```

Comment:
You need to propagate boundary conditions from the constructor into the local models, right? Should you write

`boundary_conditions = boundary_conditions`

here?

Also default boundary conditions might need to be `boundary_conditions = NamedTuple{}()`, eg an empty `NamedTuple`.

Reply:
@ali-ramadhan commented these out and said we didn't need them, but I never heard why, but what you suggest makes sense.  I will make the changes, try it again and then push the update.

Reply:
Sadly, it doesn't work when I add that line.  Maybe why it was commented out?

```
ERROR: LoadError: CUDA.jl did not successfully initialize, and is not usable.
If you did not see any other error message, try again in a new session
with the JULIA_DEBUG environment variable set to 'CUDA'.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] libcuda()
    @ CUDA ~/.julia/packages/CUDA/LTbUr/src/initialization.jl:52
  [3] macro expansion
    @ ~/.julia/packages/CUDA/LTbUr/lib/cudadrv/libcuda.jl:29 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/CUDA/LTbUr/lib/cudadrv/error.jl:94 [inlined]
  [5] cuDeviceGet
    @ ~/.julia/packages/CUDA/LTbUr/lib/utils/call.jl:26 [inlined]
  [6] CuDevice
    @ ~/.julia/packages/CUDA/LTbUr/lib/cudadrv/devices.jl:25 [inlined]
  [7] CUDA.TaskLocalState()
    @ CUDA ~/.julia/packages/CUDA/LTbUr/src/state.jl:50
  [8] task_local_state!()
    @ CUDA ~/.julia/packages/CUDA/LTbUr/src/state.jl:73
  [9] stream
    @ ~/.julia/packages/CUDA/LTbUr/src/state.jl:422 [inlined]
 [10] macro expansion
    @ ~/.julia/packages/CUDA/LTbUr/src/utilities.jl:29 [inlined]
 [11] macro expansion
    @ ~/software/Oceananigans.jl/benchmark/benchmark_shallow_water_model.jl:17 [inlined]
 [12] var"##core#260"(model#258::ShallowWaterModel{RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, CPU, Float64, UpwindBiasedFifthOrder, Nothing, NamedTuple{(:uh, :vh, :h), Tuple{typeof(Oceananigans.Forcings.zeroforcing), typeof(Oceananigans.Forcings.zeroforcing), typeof(Oceananigans.Forcings.zeroforcing)}}, Nothing, Nothing, NamedTuple{(:uh, :vh, :h), Tuple{Field{Face, Center, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}, Field{Center, Face, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}, Field{Center, Center, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}}}, NamedTuple{(), Tuple{}}, Nothing, Oceananigans.TimeSteppers.RungeKutta3TimeStepper{Float64, NamedTuple{(:uh, :vh, :h), Tuple{Field{Face, Center, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}, Field{Center, Face, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}, Field{Center, Center, Center, CPU, OffsetArrays.OffsetArray{Float64, 3, Array{Float64, 3}}, RegularRectilinearGrid{Float64, Periodic, Periodic, Flat, OffsetArrays.OffsetVector{Float64, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}}}, Float64, NamedTuple{(:x, :y, :z), Tuple{CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}, BoundaryCondition{Oceananigans.BoundaryConditions.Periodic, Nothing}}, CoordinateBoundaryConditions{Nothing, Nothing}}}}}}, Nothing}})
    @ Main ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:371
 [13] var"##sample#261"(__params::BenchmarkTools.Parameters)
    @ Main ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:377
 [14] _run(b::BenchmarkTools.Benchmark{Symbol("##benchmark#259")}, p::BenchmarkTools.Parameters; verbose::Bool, pad::String, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{4, Symbol}, NamedTuple{(:samples, :evals, :gctrial, :gcsample), Tuple{Int64, Int64, Bool, Bool}}})
    @ Main ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:405
 [15] #invokelatest#2
    @ ./essentials.jl:710 [inlined]
 [16] #run_result#37
    @ ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:32 [inlined]
 [17] run(b::BenchmarkTools.Benchmark{Symbol("##benchmark#259")}, p::BenchmarkTools.Parameters; progressid::Nothing, nleaves::Float64, ndone::Float64, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:verbose, :samples, :evals, :gctrial, :gcsample), Tuple{Bool, Int64, Int64, Bool, Bool}}})
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:94
 [18] #warmup#45
    @ ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:141 [inlined]
 [19] warmup(item::BenchmarkTools.Benchmark{Symbol("##benchmark#259")})
    @ BenchmarkTools ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:141
 [20] macro expansion
    @ ~/.julia/packages/BenchmarkTools/eCEpo/src/execution.jl:287 [inlined]
 [21] benchmark_shallow_water_model(Arch::Type{CPU}, FT::Type, N::Int64)
    @ Main ~/software/Oceananigans.jl/benchmark/benchmark_shallow_water_model.jl:16
 [22] run_benchmarks(benchmark_fun::Function; kwargs::Base.Iterators.Pairs{Symbol, Vector{T} where T, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:Architectures, :Float_types, :Ns), Tuple{Vector{DataType}, Vector{DataType}, Vector{Int64}}}})
    @ Benchmarks ~/software/Oceananigans.jl/benchmark/src/Benchmarks.jl:56
 [23] top-level scope
    @ ~/software/Oceananigans.jl/benchmark/benchmark_shallow_water_model.jl:32
in expression starting at /home/fpoulin/software/Oceananigans.jl/benchmark/benchmark_shallow_water_model.jl:32
```

Reply:
I've made a suggested change that will achieve the same thing.

Reply:
I tried removing `boundary_conditions=nothing` from the model constructor signature but then I get errors, without or without the lines commented.  But a strange one about `CUDA` not initialized.  I'm puzzled.

Reply:
@glwagner, any further advice on how you might want this changed before merging?

Reply:
If you don't propagate the boundary conditions keyword argument from the `DistributedShallowWaterModel` constructor into the local `ShallowWaterModel` constructors, isn't it impossible to provide anything but default boundary conditions in a distributed `ShallowWaterModel`?

The CUDA errors are not related to this issue, I don't think.

Reply:
I think this points to a problem with boundary conditions for at least `DistrubtedShallowWaterModel`, perhaps even `ShallowWaterModel`.  I admit that I haven't tested those very much.  

But I wonder if this is something that needs to be fixed for this PR, which focuses on fixing broken benchmarks?

Reply:
I agree it's not necessary to fix. Just seems like removing one bad line should make `DistributedShallowWater` fullly functional; until then it can't really be used for problems with non-default boundary conditions. But we can remove the line in a different PR.

Reply:
Thanks.  I will push this change and create an issue that will focus on addressing this.